### DISCUSSION TITLE
Implementing Fixed-Time Cache Revalidation in Next.js (Every 6 Hours at Specific Times)

### DISCUSSION THREAD

**Discussion by KeisukeNagakawa** [vercel/next.js#59732] [Help]
## Summary
Seeking advice on setting up a cache revalidation system in Next.js that triggers at fixed times (00:00, 06:00, 12:00, 18:00) every 6 hours. This revalidation needs to be independent of user sessions and should occur exactly at these specified times.

## Context:
I am working on a Next.js project which requires a cache revalidation process that operates on a fixed schedule. Specifically, I need the data to be revalidated every 6 hours at predetermined times (00:00, 06:00, 12:00, 18:00). This schedule should not be influenced by when a user session starts or when the page was last accessed.

## Current Understanding:
Next.js offers a revalidation feature, but it appears to be based on the duration elapsed since the last page generation, which does not align with my requirement for fixed-time revalidation.

## Questions:

1. Does Next.js have any built-in capabilities for revalidation at fixed, scheduled times, or is a custom implementation my only option?
2. If I need to create a custom solution, what would be the best practices for setting up and managing a server-side cron job within a Next.js project?
3. Are there specific challenges or potential issues I should be aware of when implementing time-based revalidation like this?

## Goal:
I'm looking for guidance and best practices from the community to implement this functionality in a way that is efficient, reliable, and scalable. Any insights, suggestions, or examples of similar implementations would be highly valuable.

**Comment by matthewraaff** [comment]
Hello,

Yes. You can use the `node-cron` package, [found here](https://www.npmjs.com/package/node-cron), and have a setup like this for once every 6 hours. There are other implementations such as [agenda](https://www.npmjs.com/package/agenda) which I think meet your requirements. 

```js
cron.schedule('0 0,6,12,18 * * *', async () => {
  .. async fetch and replace your data here
});
```

There are a few small considerations while implementing this, such as the concurrency. Make sure that your fetching logic is not ran concurrently and blocks the main thread, I'd highly recommend writing the fetch function to be async. 

Edit: [Here's what the cron does](https://crontab.guru/#0_0,6,12,18_*_*_*).

**Comment by GabrielMunguia** [comment]
Did you find any solution?

**Comment by Msantanna10** [comment]
Found a solution.
1. Create a "api/revalidate" containing the links you'd like to revalidate
2. Create an account here: https://cron-job.org/en/
3. Set up custom intervals that will trigger your "api/revalidate" endpoint

**Comment by bryannatali** [comment]
If your app is deployed to Vercel, you can use Vercel Cron Jobs, which is pretty easy to implement and I think it would solve your problem. Take a look at this: https://vercel.com/blog/cron-jobs
